alacritty/res/rect.f.glsl

```diff
@@ -1,10 +1,54 @@
 #version 330 core
 
+// We're using `origin_upper_left`, since we only known about padding from left
+// and top. If we use default `origin_bottom_left` we won't be able to offset
+// `gl_FragCoord` properly to align with the terminal grid.
+layout(origin_upper_left) in vec4 gl_FragCoord;
+
 flat in vec4 color;
 
 out vec4 FragColor;
 
+uniform int isUndercurl;
+
+uniform float cellWidth;
+uniform float cellHeight;
+uniform float paddingY;
+uniform float paddingX;
+
+uniform float undercurlThickness;
+uniform float undercurlPosition;
+
+#define PI 3.1415926538
+
 void main()
 {
-    FragColor = color;
+    if (isUndercurl == 0) {
+        FragColor = color;
+        return;
+    }
+
+    int x = int(gl_FragCoord.x - paddingX) % int(cellWidth);
+    int y = int(gl_FragCoord.y - paddingY) % int(cellHeight);
+
+    // We use `undercurlPosition` as amplitude, since it's half of descent
+    // value.
+    float undercurl = undercurlPosition / 2.
+        * cos(float(x) * 2 * PI / float(cellWidth))
+        + cellHeight - undercurlPosition;
+
+    // We subtract one from thickness since line is already one px thick.
+    float undercurl_top = undercurl + (undercurlThickness - 1) / 2.;
+    float undercurl_bottom = undercurl - (undercurlThickness - 1) / 2.;
+
+    // Assume that `y` is on the curve.
+    float alpha = 0.;
```

Comment:
This comment doesn't make sense to me? Y is an axis, why do we assume that the entire axis is on the curve?